torch/_inductor/kernel/grouped_mm_kernel_gluon.jinja

```diff
@@ -0,0 +1,457 @@
+    fence_async_shared,
+)
+
+
+# Helper class for passing arguments around partitions.
+@aggregate
+class PartitionArgs:
+    a_desc: tma.tensor_descriptor
+    b_desc: tma.tensor_descriptor
+    c_desc: tma.tensor_descriptor
+    offs_ptr: gl.tensor
+    c_ptr: gl.tensor
+    a_smem_bufs: gl.shared_memory_descriptor
+    b_smem_bufs: gl.shared_memory_descriptor
+    acc_tmem_bufs: tensor_memory_descriptor
+    load_buf_empty_bars: gl.shared_memory_descriptor
+    load_buf_full_bars: gl.shared_memory_descriptor
+    acc_buf_empty_bars: gl.shared_memory_descriptor
+    acc_buf_full_bars: gl.shared_memory_descriptor
+    G: gl.constexpr
+    M: gl.tensor
+    N: gl.tensor
+    K: gl.tensor
+    BLOCK_M: gl.constexpr
+    BLOCK_N: gl.constexpr
+    BLOCK_K: gl.constexpr
+    store_layout: gl.constexpr
+    NUM_BLOCKS: gl.constexpr
+    num_warps: gl.constexpr
+
+    def __init__(
+        self,
+        a_desc,
+        b_desc,
+        c_desc,
+        offs_ptr,
+        c_ptr,
+        a_smem_bufs,
+        b_smem_bufs,
+        acc_tmem_bufs,
+        load_buf_empty_bars,
+        load_buf_full_bars,
+        acc_buf_empty_bars,
+        acc_buf_full_bars,
+        G,
+        M,
+        N,
+        K,
+        BLOCK_M,
+        BLOCK_N,
+        BLOCK_K,
+        store_layout,
+        NUM_BLOCKS,
+        num_warps,
+    ):
+        self.a_desc = a_desc
+        self.b_desc = b_desc
+        self.c_desc = c_desc
+        self.offs_ptr = offs_ptr
+        self.c_ptr = c_ptr
+        self.a_smem_bufs = a_smem_bufs
+        self.b_smem_bufs = b_smem_bufs
+        self.acc_tmem_bufs = acc_tmem_bufs
+        self.load_buf_empty_bars = load_buf_empty_bars
+        self.load_buf_full_bars = load_buf_full_bars
+        self.acc_buf_empty_bars = acc_buf_empty_bars
+        self.acc_buf_full_bars = acc_buf_full_bars
+        self.G = G
+        self.M = M
+        self.N = N
+        self.K = K
+        self.BLOCK_M = BLOCK_M
+        self.BLOCK_N = BLOCK_N
+        self.BLOCK_K = BLOCK_K
+        self.store_layout = store_layout
+        self.NUM_BLOCKS = NUM_BLOCKS
+        self.num_warps = num_warps
+
+
+# Counter abstraction for tracking barrier index and phase.
+@aggregate
+class Counter:
+    index: gl.tensor
+    phase: gl.tensor
+    num_barriers: gl.constexpr
+
+    def __init__(self, index, phase, num_barriers):
+        self.index = index
+        self.phase = phase
+        self.num_barriers = num_barriers
+
+    @gluon.jit
+    def create(phase, num_barriers: gl.constexpr):
+        return Counter(gl.to_tensor(0), gl.to_tensor(phase), num_barriers)
+
+    @gluon.must_use_result
+    @gluon.jit
+    def next(self):
+        incr = self.index + 1
+        rollover = incr == self.num_barriers
+        index = gl.where(rollover, 0, incr)
+        phase = gl.where(rollover, self.phase ^ 1, self.phase)
+        return Counter(index, phase, self.num_barriers)
+
+
+@gluon.jit
+def load_partition(pargs: PartitionArgs):
+    tidx = gl.program_id(0)
```

Comment:
nit: (no need to fix until we decide on high level what to do), tidx typically refers to threads, program ids are bidx

Reply:
This variable is actually used as a tile index here, thus the name - this seems to be a common convention for Triton persistent kernels.